src/ruby/puppetserver-lib/puppet/server/ast_compiler.rb

```diff
@@ -5,49 +5,132 @@
 module Puppet
   module Server
     class ASTCompiler
-      def self.compile(compile_options)
+      def self.compile(compile_options, boltlib_path)
         options = compile_options['options'] || {}
 
         log_level = options['log_level']
         code = JSON.parse(compile_options['code_ast'])
 
         if options['capture_logs']
           catalog, logs = Logging.capture_logs(log_level) do
-            compile_ast(code, compile_options)
+            compile_ast(code, compile_options, boltlib_path)
           end
 
           { catalog: catalog, logs: logs }
         else
-          catalog = compile_ast(code, compile_options)
+          catalog = compile_ast(code, compile_options, boltlib_path)
           { catalog: catalog }
         end
       end
 
-      def self.compile_ast(code, compile_options)
-        Puppet[:node_name_value] = compile_options['certname']
-
-        # Use the existing environment with the requested name
-        Puppet::Pal.in_environment(compile_options['environment'],
-                                   envpath: Puppet[:environmentpath],
-                                   facts: compile_options['facts']['values'],
-                                   variables: compile_options['variables']['values']) do |pal|
-          Puppet.lookup(:pal_current_node).trusted_data = compile_options['trusted_facts']['values']
-
-          # This compiler has been configured with a node containing
-          # the requested environment, facts, and variables, and is used
-          # to compile a catalog in that context from the supplied AST.
-          pal.with_catalog_compiler do |compiler|
-            # We have to parse the AST inside the compiler block, because it
-            # initializes the necessary type loaders for us.
-            ast = Puppet::Pops::Serialization::FromDataConverter.convert(code)
-
-            compiler.evaluate(ast)
-            compiler.compile_additions
-            compiler.catalog_data_hash
+      def self.compile_ast(code, compile_options, boltlib_path)
+        # If the request requires that bolt be loaded we assume we are in a properly 
+        # configured PE environment
+        if compile_options.dig('options', 'compile_for_plan')
+          # CODEREVIEW: Where to we handle this error and what should it say?
+          unless boltlib_path
+            msg = "The path to boltlib modules must be provided in jruby-puppet boltlib-path puppetserver setting"
+            raise Puppet::Error, _(msg)
+          end
+
+          require 'bolt/apply_inventory'
+          require 'bolt/apply_target'
+          # TODO: Bolt's pal requires this so we dont have to do it in every boltib function that
+          # is explicitly forbidden in an apply block. Would it make sense to require in each function?
+          # or just do it here? 
```

Comment:
I don't understand this comment. So we don't have to do _what_ in every boltlib function? What is explicitly forbidden in an apply block?

Reply:
There are some (many) functions in boltlib that are explicitly forbidden in apply blocks. The implementation of forbidding them requires a file that is loaded by Bolt's PAL. There is not an easy way to get that change in to bolt without adding an equivalent `require` to every single function definition that is explicitly forbidden in boltlib.

Reply:
Can you update your comment to something of that^ effect? I want this to be clear for future code spelunkers.

Reply:
I will update. It should be a link to an issue. i think this is actionable. Requiring it here is just a quick workaround.